ex19/doc/WviewQV.3
'\" t
.TH "" 3 "" "Version Unknown To Mankind" "Free Widget Foundation"
.SH NAME
WviewQV
.SH DESCRIPTION

.SS "Public variables"

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
WviewQV
Name	Class	Type	Default
XtNvar	XtCVar	QVar 	0 
XtNenable	XtCEnable	QVar 	0 

.TE
.ps +2


.TP
.I "XtNvar"



.TP
.I "XtNenable"



.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
Wlabel
Name	Class	Type	Default
XtNlabel	XtCLabel	String 	NULL 
XtNcornerRoundPercent	XtCCornerRoundPercent	Int 	0 
XtNdraw_override	XtCDraw_override	XTCallbackProc 	NULL 
XtNheightIncreasePercent	XtCHeightIncreasePercent	Int 	0 
XtNleftOffsetPercent	XtCLeftOffsetPercent	Int 	5 
XtNupdate	XtCUpdate	Int 	0 

.TE
.ps +2

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
Wheel
Name	Class	Type	Default
XtNxftFont	XtCXFtFont	XftFont	"Sans-22"
XtNcallback	XtCCallback	Callback	NULL 
XtNbg_norm	XtCBg_norm	Pixel	"lightblue"
XtNbg_sel	XtCBg_sel	Pixel	"yellow"
XtNbg_hi	XtCBg_hi	Pixel	"red"
XtNfg_norm	XtCFg_norm	Pixel	"black"
XtNfg_sel	XtCFg_sel	Pixel	"green"
XtNfg_hi	XtCFg_hi	Pixel	"white"
XtNuser_data	XtCUser_data	Int 	0 
XtNfocus_group	XtCFocus_group	String 	""
XtNstate	XtCState	Int 	0 
XtNregister_focus_group	XtCRegister_focus_group	Boolean 	True 

.TE
.ps +2

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
Core
Name	Class	Type	Default
XtNx	XtCX	Position 	0 
XtNy	XtCY	Position 	0 
XtNwidth	XtCWidth	Dimension 	0 
XtNheight	XtCHeight	Dimension 	0 
borderWidth	XtCBorderWidth	Dimension 	0 
XtNcolormap	XtCColormap	Colormap 	NULL 
XtNdepth	XtCDepth	Int 	0 
destroyCallback	XtCDestroyCallback	XTCallbackList 	NULL 
XtNsensitive	XtCSensitive	Boolean 	True 
XtNtm	XtCTm	XTTMRec 	NULL 
ancestorSensitive	XtCAncestorSensitive	Boolean 	False 
accelerators	XtCAccelerators	XTTranslations 	NULL 
borderColor	XtCBorderColor	Pixel 	0 
borderPixmap	XtCBorderPixmap	Pixmap 	NULL 
background	XtCBackground	Pixel 	0 
backgroundPixmap	XtCBackgroundPixmap	Pixmap 	NULL 
mappedWhenManaged	XtCMappedWhenManaged	Boolean 	True 
XtNscreen	XtCScreen	Screen *	NULL 

.TE
.ps +2


ex19/src/WviewQV.c
/* Generated by wbuild
 * (generator version 3.3)
 */
#include <X11/IntrinsicP.h>
#include <X11/StringDefs.h>
#line 104 "WviewQV.widget"
#include "mls.h"
#line 105 "WviewQV.widget"
#include "xutil.h"
#line 106 "WviewQV.widget"
#include "micro_vars.h"
#line 107 "WviewQV.widget"
#include "converters.h"
#include <xtcw/WviewQVP.h>
static void _resolve_inheritance(
#if NeedFunctionPrototypes
WidgetClass
#endif
);
#line 15 "WviewQV.widget"
static void initialize(
#if NeedFunctionPrototypes
Widget ,Widget,ArgList ,Cardinal *
#endif
);
#line 30 "WviewQV.widget"
static Bool  cache_hit(
#if NeedFunctionPrototypes
Widget
#endif
);
#line 38 "WviewQV.widget"
static Boolean  set_values(
#if NeedFunctionPrototypes
Widget ,Widget ,Widget,ArgList ,Cardinal *
#endif
);
#line 58 "WviewQV.widget"
static void change_enable(
#if NeedFunctionPrototypes
Widget
#endif
);
#line 67 "WviewQV.widget"
static void do_update(
#if NeedFunctionPrototypes
Widget
#endif
);
#line 74 "WviewQV.widget"
static void show_data(
#if NeedFunctionPrototypes
Widget,void *,void *
#endif
);
#line 58 "WviewQV.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 58 "WviewQV.widget"
static void change_enable(Widget self)
#else
#line 58 "WviewQV.widget"
static void change_enable(self)Widget self;
#endif
#line 59 "WviewQV.widget"
{
        int remove_watcher =  ! *mv_var(((WviewQVWidget)self)->wviewQV.enable);
        TRACE(1, "%s: remove=%d", ((WviewQVWidget)self)->core.name, remove_watcher );
        mv_onwrite(((WviewQVWidget)self)->wviewQV.var, (voidfn_t)do_update, self,  remove_watcher );

        if( !remove_watcher ) do_update(self);
}
#line 67 "WviewQV.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 67 "WviewQV.widget"
static void do_update(Widget self)
#else
#line 67 "WviewQV.widget"
static void do_update(self)Widget self;
#endif
#line 68 "WviewQV.widget"
{
        TRACE(1, "%s", ((WviewQVWidget)self)->core.name );
        ((WviewQVWidget)self)->wviewQV.change_val = True;
        XtVaSetValues( self, "update", ((WviewQVWidget)self)->wlabel.update+1, NULL );
}
#line 74 "WviewQV.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 74 "WviewQV.widget"
static void show_data(Widget self,void * a,void * b)
#else
#line 74 "WviewQV.widget"
static void show_data(self,a,b)Widget self;void * a;void * b;
#endif
#line 75 "WviewQV.widget"
{

        int wh,val,h,x,y;
        char *s, str[30];
        val = *mv_var( ((WviewQVWidget)self)->wviewQV.var );
        sprintf(str,"%d", val );

        s = ((WviewQVWidget)self)->wlabel.label;
        h = ((WviewQVWidget)self)->wheel.xftFont->height;
        wh = ((WviewQVWidget)self)->core.height / 2;
        x = (((WviewQVWidget)self)->core.width - xft_textwidth( XtDisplay(self), ((WviewQVWidget)self)->wheel.xftFont, s ))/2;
        if( x < 0 ) x=0;
        y = (wh-h) / 2 + ((WviewQVWidget)self)->wheel.xftFont->ascent;
        XftDrawStringUtf8(((WviewQVWidget)self)->wlabel.draw,((WviewQVWidget)self)->wheel.xft_col + ((WviewQVWidget)self)->wheel.state, ((WviewQVWidget)self)->wheel.xftFont,
                                         x,y,(FcChar8*)s, strlen(s) );


        y += wh;
        s = str;

        x = (((WviewQVWidget)self)->core.width - xft_textwidth( XtDisplay(self), ((WviewQVWidget)self)->wheel.xftFont, s ))/2;
        if( x < 0 ) x=0;
        XftDrawStringUtf8(((WviewQVWidget)self)->wlabel.draw,((WviewQVWidget)self)->wheel.xft_col + ((WviewQVWidget)self)->wheel.state, ((WviewQVWidget)self)->wheel.xftFont,
                                         x,y,(FcChar8*)s, strlen(s) );
        ((WviewQVWidget)self)->wviewQV.drawn = val;

}

static XtResource resources[] = {
#line 7 "WviewQV.widget"
{XtNvar,XtCVar,XtRQVar,sizeof(((WviewQVRec*)NULL)->wviewQV.var),XtOffsetOf(WviewQVRec,wviewQV.var),XtRImmediate,(XtPointer)0 },
#line 8 "WviewQV.widget"
{XtNenable,XtCEnable,XtRQVar,sizeof(((WviewQVRec*)NULL)->wviewQV.enable),XtOffsetOf(WviewQVRec,wviewQV.enable),XtRImmediate,(XtPointer)0 },
};

WviewQVClassRec wviewQVClassRec = {
{ /* core_class part */
/* superclass   	*/  (WidgetClass) &wlabelClassRec,
/* class_name   	*/  "WviewQV",
/* widget_size  	*/  sizeof(WviewQVRec),
/* class_initialize 	*/  NULL,
/* class_part_initialize*/  _resolve_inheritance,
/* class_inited 	*/  FALSE,
/* initialize   	*/  initialize,
/* initialize_hook 	*/  NULL,
/* realize      	*/  XtInheritRealize,
/* actions      	*/  NULL,
/* num_actions  	*/  0,
/* resources    	*/  resources,
/* num_resources 	*/  2,
/* xrm_class    	*/  NULLQUARK,
/* compres_motion 	*/  False ,
/* compress_exposure 	*/  TRUE ,
/* compress_enterleave 	*/  False ,
/* visible_interest 	*/  False ,
/* destroy      	*/  NULL,
/* resize       	*/  XtInheritResize,
/* expose       	*/  XtInheritExpose,
/* set_values   	*/  set_values,
/* set_values_hook 	*/  NULL,
/* set_values_almost 	*/  XtInheritSetValuesAlmost,
/* get_values+hook 	*/  NULL,
/* accept_focus 	*/  XtInheritAcceptFocus,
/* version      	*/  XtVersion,
/* callback_private 	*/  NULL,
/* tm_table      	*/  NULL,
/* query_geometry 	*/  XtInheritQueryGeometry,
/* display_acceleator 	*/  XtInheritDisplayAccelerator,
/* extension    	*/  NULL 
},
{ /* Wheel_class part */
XtInherit_exec_command,
XtInherit_sig_recv,
},
{ /* Wlabel_class part */
cache_hit,
},
{ /* WviewQV_class part */
 /* dummy */  0
},
};
WidgetClass wviewQVWidgetClass = (WidgetClass) &wviewQVClassRec;
static void _resolve_inheritance(class)
WidgetClass class;
{
  WviewQVWidgetClass c __attribute__((unused)) = (WviewQVWidgetClass) class;
  WviewQVWidgetClass super __attribute__((unused));
  if (class == wviewQVWidgetClass) return;
  super = (WviewQVWidgetClass)class->core_class.superclass;
}
#line 15 "WviewQV.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 15 "WviewQV.widget"
static void initialize(Widget  request,Widget self,ArgList  args,Cardinal * num_args)
#else
#line 15 "WviewQV.widget"
static void initialize(request,self,args,num_args)Widget  request;Widget self;ArgList  args;Cardinal * num_args;
#endif
#line 16 "WviewQV.widget"
{

        ((WviewQVWidget)self)->wlabel.draw_override = show_data;
        ((WviewQVWidget)self)->core.height = (((WviewQVWidget)self)->wheel.xftFont->height) * 2.5;

        if( ((WviewQVWidget)self)->wviewQV.enable && ((WviewQVWidget)self)->wviewQV.var ) {
            mv_onwrite( ((WviewQVWidget)self)->wviewQV.enable,(voidfn_t) change_enable, self, 0 );
            change_enable( self );
        }

        ((WviewQVWidget)self)->wviewQV.change_val = False;

}
#line 30 "WviewQV.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 30 "WviewQV.widget"
static Bool  cache_hit(Widget self)
#else
#line 30 "WviewQV.widget"
static Bool  cache_hit(self)Widget self;
#endif
#line 31 "WviewQV.widget"
{
        int val = *mv_var( ((WviewQVWidget)self)->wviewQV.var );
        Bool hit = (((WviewQVWidget)self)->wviewQV.drawn == val);
        hit = hit && wlabelClassRec.wlabel_class.cache_hit(self);
        return hit;
}
#line 38 "WviewQV.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 38 "WviewQV.widget"
static Boolean  set_values(Widget  old,Widget  request,Widget self,ArgList  args,Cardinal * num_args)
#else
#line 38 "WviewQV.widget"
static Boolean  set_values(old,request,self,args,num_args)Widget  old;Widget  request;Widget self;ArgList  args;Cardinal * num_args;
#endif
#line 39 "WviewQV.widget"
{
    int i;
    for(i=0;i<*num_args;i++) {
        if( strcmp(args[i].name, "var" ) == 0 ) {
          if( ((WviewQVWidget)old)->wviewQV.var ) mv_onwrite( ((WviewQVWidget)old)->wviewQV.var,(voidfn_t) do_update,self, 1 );
          change_enable(self);
        }
        else if( strcmp(args[i].name, "enable" ) == 0 ) {
          change_enable(self);
        }
    }
    return False;
}
